File: Golang/Medium/1277_Count_Square_Submatrics_with_All_Ones/solution.go
package countsquaresubmatricswithallones

type state struct {
	row int
	col int
}

func countSquares(matrix [][]int) int {
	totalCount := 0
	for r, _ := range matrix {
		for c, _ := range matrix[0] {
			totalCount += explore(r, c, matrix, make(map[state]int))
		}
	}
	return totalCount
}

func explore(r int, c int, matrix [][]int, memo map[state]int) int {
	if !inbound(r, c, matrix) {
		return 0
	}

	key := state{r, c}
	if v, exist := memo[key]; exist {
		return v
	}

	if matrix[r][c] == 0 {
		return 0
	}

	up := explore(r-1, c, matrix, memo)
	left := explore(r, c-1, matrix, memo)
	diagonal := explore(r-1, c-1, matrix, memo)
	size := 1 + min(up, left, diagonal)

	memo[key] = size
	return size

}

func inbound(r int, c int, matrix [][]int) bool {
	rInbound := r >= 0 && r < len(matrix)
	cInbound := c >= 0 && c < len(matrix[0])
	return rInbound && cInbound
}

File: Python/Medium/1277_Count_Square_Submatrics_with_All_Ones/solution.py
class Solution:
    def countSquares(self, matrix: list[list[int]]) -> int:
        totalCount = 0
        for r in range(len(matrix)):
            for c in range(len(matrix[0])):
                totalCount += Solution.explore(r, c, matrix, {})
        return totalCount


    def explore(r, c, matrix, memo):
        if not Solution.inbound(r, c, matrix):
            return 0

        if (r,c) in memo:
            return memo[(r,c)]

        if matrix[r][c] == 0:
            return 0
        
        up = Solution.explore(r-1, c, matrix, memo)
        left = Solution.explore(r, c-1, matrix, memo)
        diagonal = Solution.explore(r-1, c-1, matrix, memo)
        size = 1 + min(up, left, diagonal)
        
        memo[(r,c)] = size
        return size

    def inbound(r, c, matrix):
        rInbound = r >= 0 and r < len(matrix)
        cInbound = c >= 0 and c < len(matrix[0])
        return rInbound and cInbound